Exercise 4.9 - Part 1:

Importing libraries.

In [26]:
# Importing libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

Creating data path.

In [27]:
# Create data path
path = r"C:\Users\Jason\Instacart Basket Analysis"

Importing customer data set.

In [28]:
# Importing customer data set
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [29]:
df_customers.shape

(206209, 10)

In [30]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [31]:
# Renaming the columns to correct names
df_customers = df_customers.rename(columns={
    'user_id' : 'User ID',
    'Surnam' : 'Surname',
    'STATE' : 'State',
    'date_joined' : 'Date Joined',
    'n_dependants' : 'Number of Dependents',
    'fam_status' : 'Family Status',
    'income' : 'Income'})


In [32]:
df_customers.head()

User ID First Name   Surname  Gender       State  Age Date Joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number of Dependents Family Status  Income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

In [33]:
# Reviewing the data for duplicates
duplicates = df_customers.duplicated()

In [34]:
# Finding the total number of duplicates
number_of_duplicates = duplicates.sum()

In [35]:
number_of_duplicates

np.int64(0)

In [36]:
# Reviewing data to check for missing values
missing_values = df_customers.isnull().sum()

In [37]:
missing_values

User ID                     0
First Name              11259
Surname                     0
Gender                      0
State                       0
Age                         0
Date Joined                 0
Number of Dependents        0
Family Status               0
Income                      0
dtype: int64

From the review, I found no duplicates, but 11259 missing values in the First Name section, but I do not believe this is an important variable for the research.

In [38]:
# Reviewing for mixed data types
df_customers.dtypes

User ID                  int64
First Name              object
Surname                 object
Gender                  object
State                   object
Age                      int64
Date Joined             object
Number of Dependents     int64
Family Status           object
Income                   int64
dtype: object

In [39]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   User ID               206209 non-null  int64 
 1   First Name            194950 non-null  object
 2   Surname               206209 non-null  object
 3   Gender                206209 non-null  object
 4   State                 206209 non-null  object
 5   Age                   206209 non-null  int64 
 6   Date Joined           206209 non-null  object
 7   Number of Dependents  206209 non-null  int64 
 8   Family Status         206209 non-null  object
 9   Income                206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [40]:
# Reviewing for mixed data types
for col in df_customers.columns:
    weird = (df_customers[[col]].map(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_customers[weird]) > 0:
        print(col)

First Name


In [41]:
# First Name possesses mixed data type
df_customers['First Name'] = df_customers['First Name'].astype('str')

In [42]:
# Reviewing again for mixed data types after changes
for col in df_customers.columns:
    weird = (df_customers[[col]].map(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_customers[weird]) > 0:
        print(col)

In [43]:
df_customers.dtypes

User ID                  int64
First Name              object
Surname                 object
Gender                  object
State                   object
Age                      int64
Date Joined             object
Number of Dependents     int64
Family Status           object
Income                   int64
dtype: object

Mixed data type changed and columns are designated to the correct data type.

6. Combine your customer data with the rest of your prepared Instacart data.

The important columns are user_id. I realised here that earlier in the exercise I changed user_id to 'User ID' for uniformity with the formatting style of the customer dataset. So first, I need to change this back to 'user_id' to enable the merge to happen.

In [44]:
# Renaming the 'User ID' column for the merge
df_customers = df_customers.rename(columns={
    'User ID':  'user_id'})

In [45]:
df_customers.head()

user_id First Name   Surname  Gender       State  Age Date Joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number of Dependents Family Status  Income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

I ran out of memory, so I needed to open a new notebook to complete the merge.

In [46]:
# Exporting the Customer dataset as CSV
df_customers.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'df_customers.csv'))